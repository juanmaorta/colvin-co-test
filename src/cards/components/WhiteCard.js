import React, { Component } from 'react'
import PropTypes from 'prop-types'
import Button from '@material-ui/core/Button'
import Card from '@material-ui/core/Card'
import CardActions from '@material-ui/core/CardActions'
import CardContent from '@material-ui/core/CardContent'
import Grid from '@material-ui/core/Grid'
import TextField from '@material-ui/core/TextField'
import Typography from '@material-ui/core/Typography'

import { withStyles } from '@material-ui/core/styles'

import styles from '../../styles'
import { ENTER_KEY, ESCAPE_KEY } from '../../util/constants'

class WhiteCard extends Component {
  constructor (props) {
    super(props)

    this.state = {
      isEditing: this.props.card.isEditing || this.props.isEditing,
      text: this.props.card.text,
      undoText: this.props.card.text
    }
  }

  onKeyDown = event => {
    switch (event.keyCode) {
      case ENTER_KEY:
        this.toggleEdit()
        break
      case ESCAPE_KEY:
        this.cancelEdit()
        break
      default:
    }
  }

  cancelEdit = () => {
    this.setState({
      isEditing: !this.state.isEditing,
      text: this.state.undoText
    })
  }

  handleChange = (event) => {
    this.setState({
      text: event.target.value
    })
  }

  toggleEdit = () => {
    if (this.state.isEditing) {
      this.props.handleEdit(
        {
          id: this.props.card.id,
          text: this.state.text,
          whiteCards: this.props.card.whiteCards,
          parentId: this.props.card.parentId
        }
      )
    }
    this.setState({
      isEditing: !this.state.isEditing,
      undoText: this.state.text
    })
  }

  render () {
    const {
      card,
      classes,
      handleDelete
    } = this.props

    return (
      <Grid item sm={6} md={4} lg={3}>
        <Card className={classes.card}>
          <CardContent>
            {!this.state.isEditing &&
              <Typography gutterBottom variant='h5' component='h2'>{card.text}</Typography>
            }
            {this.state.isEditing &&
              <TextField
                id='text'
                label='text'
                className={classes.whiteTextField}
                value={this.state.text}
                onChange={this.handleChange}
                onBlur={this.toggleEdit}
                onKeyDown={this.onKeyDown}
                margin='normal'
              />
            }
          </CardContent>
          <CardActions>
            <Button size='small' color='primary' onClick={this.toggleEdit}>
            Edit
            </Button>
            <Button size='small' color='secondary' onClick={() => handleDelete(card.id)}>
            Delete
            </Button>
          </CardActions>
        </Card>
      </Grid>
    )
  }
}

WhiteCard.propTypes = {
  classes: PropTypes.object.isRequired,
  card: PropTypes.object.isRequired,
  handleDelete: PropTypes.func.isRequired,
  handleEdit: PropTypes.func.isRequired
}

export default withStyles(styles)(WhiteCard)
