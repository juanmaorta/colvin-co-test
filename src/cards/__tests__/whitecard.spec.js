/* global expect, describe, it, jest */
import React from 'react'
import { shallow } from 'enzyme'

import WhiteCard from '../components/WhiteCard'
import * as stubs from './stubs'

describe('WhiteCard snapshots', () => {
  it('should match for WhiteCard component', () => {
    const wrapper = shallow(
      <WhiteCard
        classes={stubs.classes}
        handleDelete={jest.fn}
        handleEdit={jest.fn}
        card={stubs.card}
      />
    )
    expect(wrapper).toMatchSnapshot()
  })
})
