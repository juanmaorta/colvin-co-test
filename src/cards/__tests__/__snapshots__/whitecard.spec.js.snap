// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WhiteCard snapshots should match for WhiteCard component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <WithStyles(WhiteCard)
    card={
        Object {
            "id": 1,
            "text": "Goldfinger: He's the man, the man with...",
          }
    }
    classes={Object {}}
    handleDelete={[Function]}
    handleEdit={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "card": Object {
        "id": 1,
        "text": "Goldfinger: He's the man, the man with...",
      },
      "classes": Object {
        "appBar": "WhiteCard-appBar-1",
        "blackBackGround": "WhiteCard-blackBackGround-5",
        "blackCardContent": "WhiteCard-blackCardContent-6",
        "blackCardTitle": "WhiteCard-blackCardTitle-3",
        "buttons": "WhiteCard-buttons-2",
        "card": "WhiteCard-card-4",
        "cardGrid": "WhiteCard-cardGrid-12",
        "footer": "WhiteCard-footer-7",
        "heroButtons": "WhiteCard-heroButtons-10",
        "heroContent": "WhiteCard-heroContent-9",
        "heroUnit": "WhiteCard-heroUnit-8",
        "layout": "WhiteCard-layout-11",
      },
      "handleDelete": [Function],
      "handleEdit": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "card": Object {
          "id": 1,
          "text": "Goldfinger: He's the man, the man with...",
        },
        "classes": Object {
          "appBar": "WhiteCard-appBar-1",
          "blackBackGround": "WhiteCard-blackBackGround-5",
          "blackCardContent": "WhiteCard-blackCardContent-6",
          "blackCardTitle": "WhiteCard-blackCardTitle-3",
          "buttons": "WhiteCard-buttons-2",
          "card": "WhiteCard-card-4",
          "cardGrid": "WhiteCard-cardGrid-12",
          "footer": "WhiteCard-footer-7",
          "heroButtons": "WhiteCard-heroButtons-10",
          "heroContent": "WhiteCard-heroContent-9",
          "heroUnit": "WhiteCard-heroUnit-8",
          "layout": "WhiteCard-layout-11",
        },
        "handleDelete": [Function],
        "handleEdit": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
